### DISCUSSION TITLE
imenu functionnality (snippet inside)

### DISCUSSION THREAD

**Discussion by vindarel** [lem-project/lem#781] [Ideas]
imenu is one of the features I miss the most. 

It is true that, in CL, we can easily navigate a code base with `find-definitions` (try `C-.` on a blank line, we can choose any symbol of the image), but it isn't the same. With imenu, I can navigate effectively in the current file.

Here's a very poor man snippet:

```lisp
;; A very poor man's imenu.
(defun buffer-headings (txt)
  (loop for line in (str:lines txt)
      for parts = (str:split " " line)
      for i = 1 then (incf i)
      if (str:starts-with-p "(def" line)
      ;; collect (list line i))
        collect line))

(defun prompt-for-heading ()
  (let ((candidates (buffer-headings (buffer-text (current-buffer)))))
    (if candidates
        (prompt-for-string "Heading: "
                            :history-symbol '*imenu*
                            :completion-function (lambda (x) 
                                                   (completion-strings x candidates))
                            :test-function (lambda (name)
                                             (member name candidates :test #'string=))))))

(define-command imenu () ()
  (let ((candidate (prompt-for-heading)))
    (move-to-beginning-of-buffer)
    (search-forward (current-point) candidate)
    (message "~a" candidate)))

(define-key *global-keymap* "C-x i" 'imenu)
```

At first I wanted to grab the line numbers when parsing the file:

```lisp
(loop for line in (str:lines txt)
           for parts = (str:split " " line)
           for i = 1 then (incf i)
           if (str:starts-with-p "(def" line)
           collect (list (str:replace-all "(" "" (first parts))
                         (str:join " " (rest parts))
                         i))
```

this returns a triple, thus I don't know how to make the string completion prompt work with it.

Nor do I know how we would make the feature language agnostic.

**Comment by Sasanidas** [comment]
I think this can be really interesting to make language agnostic, specially in conjunction with https://github.com/lem-project/lem/issues/757

**Comment by Sasanidas** [comment]
Instead of using a regex, with tree-sitter we can provide an accurate tree with more layers of information

**Comment by Sasanidas** [comment]
Before using tree sitter tho, we can provide an imenu with regex (similar to what GNU Emacs or ur version does)

**Comment by Sasanidas** [comment]
https://github.com/lem-project/lem/pull/788
